### ISSUE TITLE
Clearer rules on Simple encoding

### ISSUE THREAD

**Issue by dcoutts** [cbor-wg/CBORbis#38]
The original RFC7049 is fairly ambiguous about how to treat the Simple values that are reserved or have more than one encoding (as 1 byte or 2 byte forms).

As it is, it's genuinely not clear what my decoder is supposed to accept or reject.

The original RFC says:

> As with all other major types, the 5-bit value 24 signifies a single-
> byte extension: it is followed by an additional byte to represent the
> simple value.  (To minimize confusion, only the values 32 to 255 are
> used.)

Unfortunately this does not really minimise confusion since it does not say if using the two-byte form for values 0..32 is well formed or not. It obviously suggests that you should not encode using those forms, but does not clearly say if "normal" decoders should reject them, or if "strict" decoders should. The only hint is in the errata which states:

> 4. The expression of a simple type must be as short as possible. The
>   rules for simple types follow the above rules for integers.

This is reasonable and covers the canonical case, but it suggests that it's only for the canonical case that we must avoid the two-byte form for 0..32, and by implication for normal encoders it is well formed but not recommended (just like over-long encodings for unsigned and signed numbers).

Furthermore, it is not clearly stated if the reserved simple values 24..31 are well formed or not, and if compliant decoders should reject them outright.

The current 7049bis.md document has slightly different language here. It still retains the confusing:

> As with all other major types, the 5-bit value 24 signifies a single-
> byte extension: it is followed by an additional byte to represent the
> simple value.  (To minimize confusion, only the values 32 to 255 are
> used.)

But now adds:

> An encoder MUST NOT encode False as the two-byte sequence of
> 0xf814, MUST NOT encode True as the two-byte sequence of 0xf815,
> MUST NOT encode Null as the two-byte sequence of 0xf816, and
> MUST NOT encode Undefined value as the two-byte sequence of
> 0xf817. A decoder MUST treat these two-byte sequences as an error.
> Similar prohibitions apply to the unassigned simple values as well.

So this seems to be going in the direction of declaring the range 0..32 in the two byte form to be not well-formed CBOR. But if that's the case, that would be much simpler to state explicitly.

So concretely I'd suggest:

 1. dropping the ironically confusing aside "(To minimize confusion, only the values 32 to 255 are
> used.)"
 2. replace the new "An encoder MUST NOT..." paragraph with:

```
The simple values 0..32 could in principle be encoded via the 5-bit value 24
for a single-byte extension, however this would be confusing because this range
consists of reserved simple values or values can be encoded in the single byte
form. These encodings are therefore declared to be not well-formed CBOR.
An encoder MUST NOT encode simple values 0..32 using a two-byte sequence.
A decoder MUST treat simple values 0..32 using a two-byte sequence as an error.
A corollary of this prohibition is that the reserved range of simple values are not
well-formed CBOR, and all simple values have a single well-formed encoding.
```

So this basically means that for simple values, the "normal" well-formed, and strict mode and canonical mode are the same: we only allow a single encoding for all simple values, and also ban the reserved simple values.

Alternatives are possible of course, allowing non-canonical encodings in normal mode but not strict or canonical mode. I'm not arguing for any particular choice here except that whatever is picked the language should be unambiguous.

**Comment by dcoutts** [comment]
It'd also be nice to add examples to the test vectors that are not well-formed.

**Comment by jyasskin** [comment]
#17 and #41 include an attempt to fix this, specifically at https://github.com/cbor-wg/CBORbis/pull/17/files#diff-1db0e3aacf579921911ded2cd0e6e4f5R737. How does that look to you?

**Comment by cabo** [comment]
This (except the test vectors, which went to #16) should now be fixed by f273dc1.
